File: scripts/build-gcr.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Build .gcr files for all datasets in datasets.yml
# Usage: ruby scripts/build-gcr.rb [output_dir]

require "yaml"
require "fileutils"

# Load glossarist gem
$LOAD_PATH.unshift(File.join(__dir__, "..", "..", "glossarist-ruby", "lib"))
require "glossarist"

BROWSER_ROOT = File.join(__dir__, "..")
DATASETS_DIR = File.join(BROWSER_ROOT, ".datasets")

def build_gcr(dataset_config, output_dir)
  id = dataset_config["id"]
  source_dir = File.join(DATASETS_DIR, id)
  concepts_dir = File.join(source_dir, "concepts")
  output_path = File.join(output_dir, "#{id}.gcr")

  unless File.directory?(concepts_dir)
    warn "  Skipping #{id}: no concepts/ directory"
    return
  end

  # Read register.yaml
  register_path = File.join(source_dir, "register.yaml")
  register_data = nil
  if File.exist?(register_path)
    begin
      register_data = YAML.safe_load_file(register_path, permitted_classes: [Date, Time])
    rescue Psych::SyntaxError => e
      warn "  Warning: register.yaml parse error: #{e.message}"
    end
  end

  # Read and migrate concepts
  files = Dir.glob(File.join(concepts_dir, "*.yaml"))
  puts "  Reading #{files.length} concepts..."

  concepts = []
  ref_maps = {
    ref_prefix_map: { "IEV" => "iev" },
    urn_standard_map: { "14812" => "isotc204" },
  }

  errors = 0
  files.each do |file|
    hash = YAML.safe_load_file(file, permitted_classes: [Date, Time])
    next unless hash&.dig("termid")

    # Apply schema migration (v0 → v1)
    migration = Glossarist::SchemaMigration.new(
      hash,
      from_version: "0",
      to_version: Glossarist::SchemaMigration::CURRENT_SCHEMA_VERSION,
      ref_maps: ref_maps,
    )
    migrated = migration.migrate
    concepts << migrated
  rescue => e
    errors += 1
    warn "  Error: #{File.basename(file)}: #{e.message}" if errors <= 3
  end

  puts "  Migrated #{concepts.length} concepts (#{errors} errors)"

  # Build metadata
  options = {
    title: dataset_config["title"],
    description: dataset_config["description"],
    owner: dataset_config["owner"],
    tags: dataset_config["tags"],
  }
  metadata = Glossarist::GcrMetadata.from_concepts(concepts, register_data: register_data, options: options)

  # Add register schema_version
  register_data ||= {}
  register_data["schema_version"] = Glossarist::SchemaMigration::CURRENT_SCHEMA_VERSION

  # Create .gcr
  FileUtils.mkdir_p(output_dir)
  Glossarist::GcrPackage.create(
    concepts: concepts,
    metadata: metadata,
    register_yaml: register_data,
    output_path: output_path,
  )

  puts "  Created #{output_path} (#{concepts.length} concepts)"
end

# Main
config = YAML.safe_load_file(File.join(BROWSER_ROOT, "datasets.yml"))
output_dir = ARGV[0] || File.join(BROWSER_ROOT, ".gcr")

puts "Building GCR files...\n\n"

config["datasets"].each do |ds|
  puts "#{ds['id']}:"
  build_gcr(ds, output_dir)
  puts
end

puts "Done."

File: scripts/build-gcr.mjs
#!/usr/bin/env node
/**
 * build-gcr.mjs — Build a GCR package from a harmonized dataset directory.
 *
 * Usage:
 *   node scripts/build-gcr.mjs <dataset-dir> -o <output.gcr>
 *   node scripts/build-gcr.mjs --all          # build all from datasets.yml
 *
 * The input directory must contain:
 *   - concepts/*.yaml  (harmonized concept files)
 *   - register.yaml    (optional, dataset metadata)
 *
 * The output is a ZIP archive with:
 *   - metadata.yaml    (statistics + provenance)
 *   - register.yaml    (copied from source)
 *   - concepts/*.yaml   (harmonized concept files)
 */
import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, '..');

function collectStats(conceptsDir) {
  const files = fs.readdirSync(conceptsDir).filter(f => f.endsWith('.yaml'));
  const languages = new Set();
  let withDefs = 0;
  let withSources = 0;

  for (const file of files) {
    try {
      const content = fs.readFileSync(path.join(conceptsDir, file), 'utf8');
      const concept = yaml.load(content);
      if (!concept) continue;

      const langCodes = ['eng', 'ara', 'deu', 'fra', 'spa', 'ita', 'jpn', 'kor', 'pol', 'por', 'srp', 'swe', 'zho', 'rus', 'fin', 'dan', 'nld', 'msa', 'nob', 'nno'];
      let hasDef = false;
      let hasSource = false;

      for (const lang of langCodes) {
        const lc = concept[lang];
        if (!lc) continue;
        languages.add(lang);
        if (lc.definition?.length) hasDef = true;
        if (lc.sources?.length) hasSource = true;
      }

      if (hasDef) withDefs++;
      if (hasSource) withSources++;
    } catch {}
  }

  return {
    concept_count: files.length,
    languages: [...languages].sort(),
    concepts_with_definitions: withDefs,
    concepts_with_sources: withSources,
  };
}

function buildMetadata(opts, stats) {
  const meta = {
    title: opts.title || 'Untitled Dataset',
    description: opts.description || '',
    glossarist_version: 'browser-pipeline',
    created_at: new Date().toISOString(),
    created_by: 'glossarist-vocabulary-browser build-gcr',
    statistics: stats,
    schema_version: '1.0.0',
  };

  if (opts.owner) meta.owner = opts.owner;
  if (opts.homepage) meta.homepage = opts.homepage;
  if (opts.repository) meta.repository = opts.repository;
  if (opts.tags) meta.tags = opts.tags;

  return meta;
}

function buildGcr(datasetDir, outputPath, opts = {}) {
  const conceptsDir = path.join(datasetDir, 'concepts');
  if (!fs.existsSync(conceptsDir)) {
    throw new Error(`No concepts/ directory in ${datasetDir}`);
  }

  const registerYaml = path.join(datasetDir, 'register.yaml');
  const stagingDir = path.join(ROOT, '.gcr-staging', path.basename(outputPath, '.gcr'));

  // Prepare staging
  if (fs.existsSync(stagingDir)) fs.rmSync(stagingDir, { recursive: true, force: true });
  fs.mkdirSync(path.join(stagingDir, 'concepts'), { recursive: true });

  // Collect stats
  const stats = collectStats(conceptsDir);
  console.log(`  ${stats.concept_count} concepts, ${stats.languages.length} languages`);

  // Write metadata.yaml
  const metadata = buildMetadata(opts, stats);
  fs.writeFileSync(
    path.join(stagingDir, 'metadata.yaml'),
    yaml.dump(metadata, { lineWidth: -1, noRefs: true })
  );

  // Copy register.yaml
  if (fs.existsSync(registerYaml)) {
    fs.copyFileSync(registerYaml, path.join(stagingDir, 'register.yaml'));
  }

  // Copy concept files
  const conceptFiles = fs.readdirSync(conceptsDir).filter(f => f.endsWith('.yaml'));
  for (const f of conceptFiles) {
    fs.copyFileSync(path.join(conceptsDir, f), path.join(stagingDir, 'concepts', f));
  }

  // Build ZIP
  const outputDir = path.dirname(outputPath);
  if (!fs.existsSync(outputDir)) fs.mkdirSync(outputDir, { recursive: true });
  if (fs.existsSync(outputPath)) fs.unlinkSync(outputPath);

  execSync(`cd "${stagingDir}" && zip -r -q "${path.resolve(outputPath)}" .`, { stdio: 'pipe' });

  // Cleanup staging
  fs.rmSync(stagingDir, { recursive: true, force: true });

  const size = fs.statSync(outputPath).size;
  console.log(`  Built ${outputPath} (${(size / 1024).toFixed(0)} KB)`);
}

// --- CLI ---
const args = process.argv.slice(2);
const allMode = args.includes('--all');

if (allMode) {
  // Build all datasets from datasets.yml
  const configPath = path.join(ROOT, 'datasets.yml');
  const config = yaml.load(fs.readFileSync(configPath, 'utf8'));
  const datasetsDir = path.join(ROOT, '.datasets');
  const gcrDir = path.join(ROOT, '.gcr');

  for (const ds of config.datasets) {
    const datasetDir = path.join(datasetsDir, ds.id);
    const outputPath = path.join(gcrDir, `${ds.id}.gcr`);

    if (!fs.existsSync(path.join(datasetDir, 'concepts'))) {
      console.log(`${ds.id}: no concepts/ directory, skipping`);
      continue;
    }

    console.log(`${ds.id}:`);
    buildGcr(datasetDir, outputPath, {
      title: ds.title,
      description: ds.description,
      owner: ds.owner,
      homepage: ds.existingSiteUrl,
      repository: ds.sourceRepo,
      tags: ds.tags,
    });
  }
} else {
  // Build single dataset
  let datasetDir = null;
  let outputPath = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '-o' && args[i + 1]) {
      outputPath = args[++i];
    } else if (!datasetDir) {
      datasetDir = args[i];
    }
  }

  if (!datasetDir || !outputPath) {
    console.error('Usage: node scripts/build-gcr.mjs <dataset-dir> -o <output.gcr>');
    console.error('       node scripts/build-gcr.mjs --all');
    process.exit(1);
  }

  buildGcr(datasetDir, outputPath);
}

console.log('\nDone.');
